app/services/tunnel.py

```diff
@@ -131,26 +144,37 @@ def get_tunnel_details(self, result: dict, job_id: str) -> Tuple[str, str]:
         ip_address = next(x["Address"] for x in nodes if x["ID"] in allocation_node)
         allocated_ports = values(allocation_info, "Resources/Networks/0/DynamicPorts/*")
         ssh_port = next(x for x in allocated_ports if x["Label"] == "ssh")["Value"]
-
         if current_app.config["ENV"] == "development":
             ip_address = current_app.config["SEA_HOST"]
 
         return (ssh_port, ip_address)
 
+    def get_tcp_port(self) -> int:
+        try:
+            new_port = redis_client.spop("open_tcp_ports")
+            return int(new_port)
+        except redis.exceptions.ConnectionError:
```

Comment:
there's no need to subclass this error - especially since we're catching *everything* in the next stanza and calling that a "no ports available" failure.  I think we either need to tighten up the "no ports available" condition or just raise everything through.